### PR TITLE
[phoneNumberPrivacy] add ratelimiting logic with db connection [unit tests in progress]

### PR DESCRIPTION
**Pull request by aslawson**
### Description

- enforces ratelimiting on phone number salt generation
- query quota is calculated via: unverifiedQueryCount + verifiedQueryCount + (queryPerTransaction * transactionCount)
- number of queries already performed is stored in the `accounts` table in the DB as `num_lookups`
- on request, the `num_lookups` is incremented
- **on failure of db table right now, the error is just logged and the salt would still be returned** -- up for discussion on how we want to handle this

- currently pointing to `https://alfajores-forno.celo-testnet.org` for attestation contract interaction and `https://alfajores-blockscout.celo-testnet.org` for transaction counts.  this may change if we determine these services do not have high enough reliability for our use case.

### Tested

Heavily tested manually.  Unit Tests in progress

`curl -v "http://localhost:5000/celo-phone-number-privacy/us-central1/getSalt" -d '{"queryPhoneNumber": "5555555555", "account":"0x117ea45d497ab022b85494ba3ab6f52969bf6831", "phoneNumber":"+5555555555"}' -H 'Content-Type: application/json'`

### Other changes

NA

### Related issues

- Fixes https://github.com/celo-org/celo-labs/issues/330

### Backwards compatibility

NA -- Service not used yet

### CODE REVIEW

File: packages/phone-number-privacy/package.json
Comment (jmrossy) [inline packages/phone-number-privacy/package.json]: We can only use published versions in cloud functions, so this needs to be 0.3.1

File: packages/phone-number-privacy/src/config.ts
Comment (jmrossy) [inline packages/phone-number-privacy/src/config.ts]: These looks right to me for now :)

File: packages/phone-number-privacy/src/config.ts
Comment (jmrossy) [inline packages/phone-number-privacy/src/config.ts]: Consider renaming to `unverifiedQueryMax` and ditto below

File: packages/phone-number-privacy/src/web3/contracts.ts
Comment (jmrossy) [inline packages/phone-number-privacy/src/web3/contracts.ts]: Given that we don't explicitly declare web3 as a dep of this project, it'd be better to let CK handle the interactions with web. Consider using `newKit` instead of `newKitFromWeb3`

File: packages/phone-number-privacy/src/index.ts
Comment (jmrossy) [inline packages/phone-number-privacy/src/index.ts]: Nit: I'd recommend we move all code out of index.ts except the API endpoints, which is a common structure for cloud functions I believe

File: packages/phone-number-privacy/src/index.ts
Comment (jmrossy) [inline packages/phone-number-privacy/src/index.ts]: Don't handle the response directly here. Just let this function worry about BLS and then have the controllers layer (getSalt above) worry about dealing with HTTP codes and such.

File: packages/phone-number-privacy/src/index.ts
Comment (jmrossy) [inline packages/phone-number-privacy/src/index.ts]: Any reason we're using raw promises here instead of making the getSalt handler async?

File: packages/phone-number-privacy/src/index.ts
Comment (jmrossy) [inline packages/phone-number-privacy/src/index.ts]: ditto

**Review by jmrossy — COMMENTED** [review]
Great progress! See comments below :)
Overall looks good but I think we could structure the code better to avoid having everything live in index.ts. Maybe you were already planning to do that in a follow-up PR, not sure.

File: packages/phone-number-privacy/src/index.ts
Comment (aslawson) [reply packages/phone-number-privacy/src/index.ts]: good point -- I was just trying to get something functional out first, but starting to refactor may make the tests cleaner as well

File: packages/phone-number-privacy/src/index.ts
Comment (aslawson) [reply packages/phone-number-privacy/src/index.ts]: I blame it on my lack of experience with typescript & firebase.  I didn't think it was possible with the firebase functions, but looks like it is
https://www.youtube.com/watch?v=Jr7pDZ1RAUg

I'll update -- that will clean up the code a lot

File: packages/phone-number-privacy/src/config.ts
Comment (codyborn) [inline packages/phone-number-privacy/src/config.ts]: Without looking at the code, I would assume that if the user is verified, we would go with the verifiedQueryCount and not (unverifiedQueryCount + verifiedQueryCount). We can rename it to "additionalVerifiedQueryCount" or remove the +unverifiedQueryCount to make it more intuitive.

File: packages/phone-number-privacy/src/index.ts
Comment (codyborn) [inline packages/phone-number-privacy/src/index.ts]: This logic is going to grow quickly. We should create some classes to make this more extensible in the future. For example, each of these queryQuota changes can be generalized as a `QuotaRule` and we can use `QuotaRuleAggregator` to call each rule and aggregate them. Today we have 3 rules (unverified, verified, txCount). Each of these rules can be calculated in parallel and all add some number to the end queryQuota. I imagine we'll want to add more `QuotaRule` types in the future, but the aggregation logic shouldn't change.

File: packages/phone-number-privacy/src/index.ts
Comment (codyborn) [inline packages/phone-number-privacy/src/index.ts]: What is the max size of this response? We'll need to paginate processing if this grows linearly with number of txs.

File: packages/phone-number-privacy/src/index.ts
Comment (codyborn) [inline packages/phone-number-privacy/src/index.ts]: Using blockscout doesn't feel like a good long term solution. I'm concerned about SLA and security of the blockscout instance. The more services that we put in our critical path, the lower our SLA will be. I think a better short-term approach is to query the latest nonce for a given account from the full-node that we're running.

File: packages/phone-number-privacy/src/index.ts
Comment (codyborn) [inline packages/phone-number-privacy/src/index.ts]: AFAIK we're planning on using Prometheus and Stackdriver across all CLabs services for monitoring. Does it make sense to integrate now so we can start logging these errors?

File: packages/phone-number-privacy/src/index.ts
Comment (codyborn) [inline packages/phone-number-privacy/src/index.ts]: We should create a new class for wrapping the knex object and performing these high level operations. The more internals we expose, the harder it will be to guarantee our assumptions. Having a separate class will also make it easier to mock during testing. I recommend creating a new class (ex. `QueryStateStore`) that exposes `getPerformedQueryCount` and `incrementQueryCount`.

File: packages/phone-number-privacy/src/index.ts
Comment (codyborn) [inline packages/phone-number-privacy/src/index.ts]: The query and increment logic should be atomic. For example, if I hit two servers simultaneously, both query my count as 32 and increment my count by 1, then my count will reach 34, which breaks the rules.

Some ideas to fix this:
1. Use serializable transactions. This will ensure that only one thread can read and modify the tx at a time. The second update will fail which is desired in our scenario.
See "SERIALIZABLE transactions" [here](https://www.2ndquadrant.com/en/blog/postgresql-anti-patterns-read-modify-write-cycles/).
2. Pass in the read value ('expectedLookups') into incrementQueryCount and fail if it's different than expected.
```
knex(ACCOUNTS_TABLE)
      .where(ACCOUNTS_COLUMNS.address, account)
      .where(ACCOUNTS_COLUMNS.numLookups, expectedLookups)
      .increment(ACCOUNTS_COLUMNS.numLookups, 1)
```
See "Optimistic concurrency controls" in the linked article.
3. Increment first, then read
This has the negative side-effect of potentially blocking a user too soon. However, if a user is spamming us perhaps they deserve to be blocked 😝

File: packages/phone-number-privacy/src/index.ts
Comment (codyborn) [inline packages/phone-number-privacy/src/index.ts]: We should perform thorough validation of any content that's passed to us from an untrusted client.

File: packages/phone-number-privacy/src/index.ts
Comment (codyborn) [inline packages/phone-number-privacy/src/index.ts]: Should this be configurable?

File: packages/phone-number-privacy/src/index.ts
Comment (jmrossy) [reply packages/phone-number-privacy/src/index.ts]: These logs do go to stackdriver. Probably makes sense to hold off on any further work on logging + monitoring til we have the final infra setup planned out.

File: packages/phone-number-privacy/src/index.ts
Comment (jmrossy) [reply packages/phone-number-privacy/src/index.ts]: Users can set arbitrary nonces in txs so long as it's greater than the previous one, right? . I think one way or another we need a reliable count on sent txs. Lets discuss it in our next sync

File: packages/phone-number-privacy/src/config.ts
Comment (aslawson) [reply packages/phone-number-privacy/src/config.ts]: sure -- sounds good

File: packages/phone-number-privacy/src/index.ts
Comment (aslawson) [reply packages/phone-number-privacy/src/index.ts]: interesting -- I was not aware that nonce can be set arbitrarily

File: packages/phone-number-privacy/src/index.ts
Comment (aslawson) [reply packages/phone-number-privacy/src/index.ts]: its not in the mobile application, and I don't expect it to change much.

File: packages/phone-number-privacy/src/index.ts
Comment (aslawson) [reply packages/phone-number-privacy/src/index.ts]: agreed

File: packages/phone-number-privacy/src/index.ts
Comment (codyborn) [reply packages/phone-number-privacy/src/index.ts]: A user can set an arbitrary nonce, but the tx won't get mined unless it's +1 from the previous nonce. This is to ensure that someone doesn't censor a tx and lead you to taking actions that you didn't intend. There are two ways that we can query the RPC API for next nonce. One method for only taking into account on-chain txs and another method uses nonces in the mem-pool. For the latter method, I'm not sure if there's validation that the txs are strictly in order (non-skipping), however the former method definitely does.

File: packages/phone-number-privacy/src/index.ts
Comment (aslawson) [reply packages/phone-number-privacy/src/index.ts]: this is a good point.
1. it seems we can pass this setting in via a raw query.  it says it lasts the current session (which I would think it just the full firebase request that creates the connection to the DB)
https://stackoverflow.com/questions/28181998/knex-js-postgresql-how-to-change-transaction-isolation-level
2. the only complication here with knex is that i'm already doing an update OR create, so if the update fails, its assuming its because the entry doesn't exist.  if I add this additional condition, it will also try to perform the create (which would also fail since it exists, but its just wasted requests)
3. ya -- we could do this and would only be an issue if there's an error later that prevents the user from getting the salt (we've essentially wasted their query)

i'm not sure how to test -- particularly the first one would be hard to tell if its been implemented correctly

Comment (aslawson) [reply packages/phone-number-privacy/src/index.ts]: I added a sleep to simulate a race condition.  With adding
`  await knex.raw('BEGIN TRANSACTION ISOLATION LEVEL SERIALIZABLE;')`

it too also will fail on the update and then try the insert.  there really is no way to tell the difference between a failed update due to an entry nonexistence vs rejection to race condition.  the db returns just "0" to represent 0 rows updated rather than throw a parsable error :/

oddly the second request to "insert" then hangs and requires adding a timeout.

lets discuss in our next meeting

File: packages/phone-number-privacy/src/index.ts
Comment (aslawson) [reply packages/phone-number-privacy/src/index.ts]: I like this idea as an improvement.  I'm going to leave as is for now due to time constraints and given that the current rules aren't exclusive -- we don't need to get all transactions if the user isn't even verified.

File: packages/phone-number-privacy/src/index.ts
Comment (aslawson) [reply packages/phone-number-privacy/src/index.ts]: good point -- i'll wait until we discuss removing blockscout

File: packages/phone-number-privacy/src/common/identity.ts
Comment (jmrossy) [inline packages/phone-number-privacy/src/common/identity.ts]: `authenticateUser` may be a more descriptive and common name for this

File: packages/phone-number-privacy/src/database/database.ts
Comment (jmrossy) [inline packages/phone-number-privacy/src/database/database.ts]: what does this do?

File: packages/phone-number-privacy/src/database/database.ts
Comment (jmrossy) [inline packages/phone-number-privacy/src/database/database.ts]: Maybe move this function here now that we have a dedicated database file :)

File: packages/phone-number-privacy/src/index.ts
Comment (jmrossy) [inline packages/phone-number-privacy/src/index.ts]: Consider moving to config.ts

File: packages/phone-number-privacy/src/index.ts
Comment (jmrossy) [inline packages/phone-number-privacy/src/index.ts:25]: Should be no need for a `.catch` here given the outer try/catch in this function

File: packages/phone-number-privacy/src/salt-generation/query-quota.ts
Comment (jmrossy) [inline packages/phone-number-privacy/src/salt-generation/query-quota.ts]: ditto, move to config and don't duplicate it

File: packages/phone-number-privacy/src/salt-generation/query-quota.ts
Comment (jmrossy) [inline packages/phone-number-privacy/src/salt-generation/query-quota.ts]: why create a module-scoped CK var here instead of calling getCK() where you need it?

File: packages/phone-number-privacy/src/salt-generation/query-quota.ts
Comment (jmrossy) [inline packages/phone-number-privacy/src/salt-generation/query-quota.ts]: Similar to index, lets not mix promises and async/await.
If you're having to add the .catch to make the linter happy it means something else is wrong. `await` should already unwrap the promise

File: packages/phone-number-privacy/src/salt-generation/query-quota.ts
Comment (jmrossy) [inline packages/phone-number-privacy/src/salt-generation/query-quota.ts]: nit, no need for explicit `: number` typing here right? TS should already know it

File: packages/phone-number-privacy/src/salt-generation/query-quota.ts
Comment (jmrossy) [inline packages/phone-number-privacy/src/salt-generation/query-quota.ts]: maybe we should `throw here`. Not an urgent concern at all and we can make a pass at good error handling throughout the app in a later PR

**Review by jmrossy — COMMENTED** [review]
Almost ready to merge! See comments below :)

File: packages/phone-number-privacy/src/salt-generation/query-quota.ts
Comment (aslawson) [reply packages/phone-number-privacy/src/salt-generation/query-quota.ts]: oh it was unused in the other location... but i'll put in config anyway

File: packages/phone-number-privacy/src/index.ts
Comment (aslawson) [reply packages/phone-number-privacy/src/index.ts:25]: I suppose -- I was leaving it open in case we wanted to handle this differently.  right now if the db request fails we allow the query to be made.  if i allow the try/catch to be used it would return a 500.  it depends what we how we want to handle these errors

File: packages/phone-number-privacy/src/database/database.ts
Comment (aslawson) [reply packages/phone-number-privacy/src/database/database.ts]: handle concurrent requests (see discussion with @codyborn above)

"This level emulates serial transaction execution, as if transactions had been executed one after another, serially, rather than concurrently."

https://www.postgresql.org/docs/current/sql-set-transaction.html
^ i says it should send an error, although in practice I just see it fail on update and return a 0 (as in 0 rows updated)

File: packages/phone-number-privacy/src/salt-generation/query-quota.ts
Comment (aslawson) [reply packages/phone-number-privacy/src/salt-generation/query-quota.ts]: copy pasta in the refactor -- will update

File: packages/phone-number-privacy/src/salt-generation/query-quota.ts
Comment (aslawson) [reply packages/phone-number-privacy/src/salt-generation/query-quota.ts]: i was originally part of the db request handling .then()/.catch() -- but I didn't want to handle it in wrapper of the account table itself since I felt error handling for the db request should be handled on a per-case basis.  what we want to do when there is a db error is circumstantial.

here it catches it and allows the transaction

File: packages/phone-number-privacy/src/salt-generation/query-quota.ts
Comment (aslawson) [reply packages/phone-number-privacy/src/salt-generation/query-quota.ts]: same thing -- it depends really if we want to refuse the user to query if there is a failure on our end.  if we want to avoid the potential of being a bottleneck to transactions, we'd want to error on allowing.  but if the risk for security is too high then we'd want to refuse the request

File: packages/phone-number-privacy/src/index.ts
Comment (aslawson) [reply packages/phone-number-privacy/src/index.ts]: removed blockscout

File: packages/phone-number-privacy/src/index.ts
Comment (aslawson) [reply packages/phone-number-privacy/src/index.ts]: removed blockscout

File: packages/phone-number-privacy/src/salt-generation/query-quota.ts
Comment (aslawson) [reply packages/phone-number-privacy/src/salt-generation/query-quota.ts]: removed

File: packages/phone-number-privacy/src/database/database.ts
Comment (jmrossy) [inline packages/phone-number-privacy/src/database/database.ts]: Nit, no need for the connectToDatabase wrapper function at this point, may as well just assign this to the `knex`object.
If we find later that we should defer the knex object creation we can initiate it in getDatabase()

File: packages/phone-number-privacy/test/index.test.ts
Comment (jmrossy) [inline packages/phone-number-privacy/test/index.test.ts:22]: Nit a more common form for this would be  would be 
`jest.fn( () => {new Promise((resolve) => resolve())}),`

Also why the promise here? Keep in mind `await` can used on non-promises, Js handles it fine :)

File: packages/phone-number-privacy/test/salt-generation/query-quota.test.ts
Comment (jmrossy) [inline packages/phone-number-privacy/test/salt-generation/query-quota.test.ts:12]: why do you import utils here via require instead of just `import`ing it at the top?

File: packages/phone-number-privacy/test/salt-generation/query-quota.test.ts
Comment (jmrossy) [inline packages/phone-number-privacy/test/salt-generation/query-quota.test.ts]: Maybe clean up the comments here a bit unless we need these?

**Review by jmrossy — APPROVED** [review]
Looks good! A few small comments but nothing major. 
Let's 🚢 it and keep building

File: packages/phone-number-privacy/test/salt-generation/query-quota.test.ts
Comment (aslawson) [reply packages/phone-number-privacy/test/salt-generation/query-quota.test.ts:12]: jest cannot find it otherwise...  apparently anything within the mock context needs to be defined there

File: packages/phone-number-privacy/test/index.test.ts
Comment (aslawson) [reply packages/phone-number-privacy/test/index.test.ts:22]: RE: promises -- i didn't find that to always be the case -- if I had a .catch() after the promise call, then jest would complain that just passing the value back didn't have a .catch() method

File: packages/phone-number-privacy/test/salt-generation/query-quota.test.ts
Comment (aslawson) [reply packages/phone-number-privacy/test/salt-generation/query-quota.test.ts]: k

File: packages/phone-number-privacy/src/database/database.ts
Comment (aslawson) [reply packages/phone-number-privacy/src/database/database.ts]: kk

File: packages/phone-number-privacy/test/salt-generation/query-quota.test.ts
Comment (jmrossy) [reply packages/phone-number-privacy/test/salt-generation/query-quota.test.ts:12]: Hmm, maybe because of how jest implicitly hoists all mocks to the top of the file, except for variables that contain 'mock' in the name (IIRC)

**Pull request merged**
